src/omv/ports/stm32/jpeg.c

```diff
@@ -0,0 +1,713 @@
+                    temp, JPEG_OUTPUT_CHUNK_SIZE, JPEG_CODEC_TIMEOUT);
+
+    // Set handles for full decoding.
+    HAL_JPEG_RegisterGetDataCallback(&JPEG_state.jpeg_descr, jpeg_decompress_get_data);
+    HAL_JPEG_RegisterDataReadyCallback(&JPEG_state.jpeg_descr, jpeg_decompress_data_ready);
+
+    int mcu_w = JPEG_MCU_W;
+    int mcu_h = JPEG_MCU_H;
+    int mcu_size = JPEG_444_GS_MCU_SIZE;
+    DMA2D_HandleTypeDef DMA2D_Handle = {};
+
+    if (JPEG_state.jpeg_descr.Conf.ColorSpace == JPEG_YCBCR_COLORSPACE) {
+        switch (JPEG_state.jpeg_descr.Conf.ChromaSubsampling) {
+            case JPEG_444_SUBSAMPLING: {
+                mcu_w = JPEG_MCU_W;
+                mcu_h = JPEG_MCU_H;
+                mcu_size = JPEG_444_YCBCR_MCU_SIZE;
+                DMA2D_Handle.LayerCfg[1].ChromaSubSampling = DMA2D_NO_CSS;
+                break;
+            }
+            case JPEG_420_SUBSAMPLING: {
+                mcu_w = JPEG_MCU_W * 2;
+                mcu_h = JPEG_MCU_H * 2;
+                mcu_size = JPEG_420_YCBCR_MCU_SIZE;
+                DMA2D_Handle.LayerCfg[1].ChromaSubSampling = DMA2D_CSS_420;
+                break;
+            }
+            case JPEG_422_SUBSAMPLING: {
+                mcu_w = JPEG_MCU_W * 2;
+                mcu_h = JPEG_MCU_H;
+                mcu_size = JPEG_422_YCBCR_MCU_SIZE;
+                DMA2D_Handle.LayerCfg[1].ChromaSubSampling = DMA2D_CSS_422;
+                break;
+            }
+            default: {
+                break;
+            }
+        }
+
+        if (dst->is_color) {
+            DMA2D_Handle.Instance = DMA2D;
+
+            // Configure DMA2D output.
+            DMA2D_Handle.Init.Mode = DMA2D_M2M_PFC;
+            DMA2D_Handle.Init.ColorMode = DMA2D_OUTPUT_RGB565;
+            DMA2D_Handle.Init.OutputOffset = 0;
+            DMA2D_Handle.Init.AlphaInverted = DMA2D_REGULAR_ALPHA;
+            DMA2D_Handle.Init.RedBlueSwap = DMA2D_RB_REGULAR;
+            DMA2D_Handle.Init.BytesSwap = DMA2D_BYTES_REGULAR;
+            DMA2D_Handle.Init.LineOffsetMode = DMA2D_LOM_PIXELS;
+
+            // Configure DMA2D input.
+            DMA2D_Handle.LayerCfg[1].InputOffset = (((src->w + mcu_w - 1) / mcu_w) * mcu_w) - src->w;
+            DMA2D_Handle.LayerCfg[1].InputColorMode = DMA2D_INPUT_YCBCR;
+            DMA2D_Handle.LayerCfg[1].AlphaMode = DMA2D_NO_MODIF_ALPHA;
+            DMA2D_Handle.LayerCfg[1].InputAlpha = 0xFF;
+            DMA2D_Handle.LayerCfg[1].AlphaInverted = DMA2D_REGULAR_ALPHA;
+            DMA2D_Handle.LayerCfg[1].RedBlueSwap = DMA2D_RB_REGULAR;
+
+            // DMA2D initialization.
+            HAL_DMA2D_Init(&DMA2D_Handle);
+            HAL_DMA2D_ConfigLayer(&DMA2D_Handle, 1);
+
+            // Invalidate the dst image for DMA2D.
+            SCB_InvalidateDCache_by_Addr((uint32_t *) dst->data, image_size(dst));
+        }
+    } else if (JPEG_state.jpeg_descr.Conf.ColorSpace == JPEG_CMYK_COLORSPACE) {
+        fb_free(); // JPEG_state.jpeg_descr.pJpegInBuffPtr
```

Comment:
But now you have to check if it was allocated or aligned, maybe `JPEG_state.jpeg_descr.pJpegInBuffPtr != src->data`

Reply:
Or `if (((uint32_t) src->data) % __SCB_DCACHE_LINE_SIZE) {` like below.

Reply:
Good catch.

Reply:
Fixed